### PR TITLE
ENT-2509 - Make @InitiatedBy flows overridable via node config

### PR DESCRIPTION
**Pull request by roastario**
At the moment, corda allows you to specify at most one flow with @InitiatedBy as a responder to a flow. 

In production, it is likely that a single cordapp with contain a "base" responder, which other users of the dapp will want to configure for use with their backends. 

This can be achieved either by subclassing a flow, or by providing an alternative implementation. 

This PR will provide the following functionality

1. Allow users to subclass flows, and deterministically know that this subclassed implementation will be used as the responder
2. Allow users to specify a flow to respond to the Initiator regardless of it's place in the inheritance tree. And know deterministically that this implementation will be used as the responder. 

The overriding will be controlled via the node.conf with a new section:

    flowOverrides {
        overrides=[
            {
                initiator="net.corda.flows.overrides.Initiator"
                responder="net.corda.flows.overrides.LegacyPostgresResponder"
            }
        ]
    }

### CODE REVIEW

**Comment by roger3cev** [comment]
This looks promising - is there a design doc or some high level notes I can read? Cheers!

**Comment by mikehearn** [comment]
What is a flowManager and can we not open PRs that aren't intended to be reviewed? Thanks! :)

**Comment by mikehearn** [comment]
Is editing the config file the way to go here? Could we maybe take the overrides from annotations in the app itself, to convert the common case of one-app-one-customisation-app into simply installing an app, rather than install+reconfigure? Less to go wrong this way, less potential skew between dev/test/prod environments, etc.

I guess the config can remain for the hopefully rare case where two apps are fighting over the same flow and you have to pick a winner.

**Comment by mikehearn** [comment]
Builds are failing and it seems to altering public API ....

**Comment by mikehearn** [comment]
Ah, interesting, I hadn't picked up on the distance trick. Very nice!

File: core/src/test/kotlin/net/corda/core/flows/ReceiveAllFlowTests.kt [hunk 1]
```diff
@@ -2,14 +2,13 @@ package net.corda.core.flows
```
Comment (mikehearn) [inline core/src/test/kotlin/net/corda/core/flows/ReceiveAllFlowTests.kt]: I don't understand the changes in this file - doesn't look related? Should this file be in a separate PR?

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+import net.corda.core.utilities.contextLogger
+import net.corda.core.utilities.debug
+import net.corda.node.internal.classloading.requireAnnotation
+import net.corda.node.services.config.FlowOverrideConfig
+import net.corda.node.services.statemachine.appName
+import net.corda.node.services.statemachine.flowVersionAndInitiatingClass
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.reflect.KClass
+
+interface FlowManager {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: Remove blank lines, add some comments. This file is the core of the whole change but I could really benefit from some explanations throughout the code.

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    fun <F : FlowLogic<*>> registerInitiatedFlow(responder: Class<F>)
+
+    fun getFlowFactoryForInitiatingFlow(initiatedFlowClass: Class<out FlowLogic<*>>): InitiatedFlowFactory<*>?
+
+    fun validateRegistrations()
+}
+
+
+open class NodeFlowManager(flowOverrides: FlowOverrideConfig? = null) : FlowManager {
+
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: blank lines

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+    val flowOverrides = (flowOverrides
+            ?: FlowOverrideConfig()).overrides.map { it.initiator to it.responder }.toMutableMap()
+
+    companion object {
+        private val log = contextLogger()
+    }
+
+    private val flowFactories = ConcurrentHashMap<Class<out FlowLogic<*>>, MutableList<RegisteredFlowContainer>>()
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: Is this class thread safe? If so should it be annotated with `@ThreadSafe`?

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        registerInitiatedFlowFactory(initiator, flowFactory, responder)
+        log.info("Registered ${initiator.name} to initiate ${responder.name} (version $version)")
+    }
+
+
+    private fun <F : FlowLogic<*>> registerInitiatedFlowFactory(initiatingFlowClass: Class<out FlowLogic<*>>,
+                                                                flowFactory: InitiatedFlowFactory<F>,
+                                                                initiatedFlowClass: Class<F>?) {
+
+        check(flowFactory !is InitiatedFlowFactory.Core) { "This should only be used for CORDAPP flows" }
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: CORDAPP -> CordApp.

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        if (listOfFlowsForInitiator.isNotEmpty() && listOfFlowsForInitiator.first().type == FlowType.CORE) {
+            throw IllegalStateException("Attempting to register over an existing platform flow: $initiatingFlowClass")
+        }
+        synchronized(listOfFlowsForInitiator) {
+            val flowToAdd = RegisteredFlowContainer(initiatingFlowClass, initiatedFlowClass, flowFactory, FlowType.CORDAPP)
+            val flowWeightComparator = FlowWeightComparator(initiatingFlowClass, flowOverrides)
+            listOfFlowsForInitiator.add(flowToAdd)
+            listOfFlowsForInitiator.sortWith(flowWeightComparator)
+            if (listOfFlowsForInitiator.size > 1) {
+                log.info("Mutiliple flows are registered for InitiatingFlow: $initiatingFlowClass, currently using: ${listOfFlowsForInitiator.first().initiatedFlowClass}")
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: typo: Multiple

Should this be a warning rather than a log.info?

File: samples/flow-override-demo/build.gradle [hunk 7]
```diff
@@ -0,0 +1,70 @@
+apply plugin: 'kotlin'
```
Comment (mikehearn) [inline samples/flow-override-demo/build.gradle]: Rather than a new sample to test, could this be integrated with an existing one like the trader demo? I realise it may be a bit "forced" scenario-wise but it'd be nice not to have a dedicated sample for it.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt [hunk 8]
```diff
@@ -156,7 +155,7 @@ class StartedMockNode private constructor(private val node: TestStartedNode) {
     fun <T> startFlow(logic: FlowLogic<T>): CordaFuture<T> = node.services.startFlow(logic, node.services.newContext()).getOrThrow().resultFuture
 
     /** Register a flow that is initiated by another flow .**/
-    fun <F : FlowLogic<*>> registerInitiatedFlow(initiatedFlowClass: Class<F>): Observable<F> = node.registerInitiatedFlow(initiatedFlowClass)
+    fun <F : FlowLogic<*>> registerInitiatedFlow(initiatedFlowClass: Class<F>) = node.registerInitiatedFlow(initiatedFlowClass)
```
Comment (mikehearn) [inline testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt]: This is public API so we shouldn't be *removing* explicit return types, if anything we need more of them!

File: core/src/test/kotlin/net/corda/core/flows/ReceiveAllFlowTests.kt (hunk 1 above)
Comment (roastario) [reply core/src/test/kotlin/net/corda/core/flows/ReceiveAllFlowTests.kt]: yep, I've undone these changes

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt (hunk 2 above)
Comment (roastario) [reply node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: done

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt (hunk 3 above)
Comment (roastario) [reply node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: done

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt (hunk 2 above)
Comment (mikehearn) [reply node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: Thanks. Maybe use standard kdoc comments `/** lines starting with asterisk */`

💡 IntelliJ column select mode is super useful for this. Shift-Cmd-8, shift arrow cursors to create a large vertical select bar, press * once and it's inserted on all the lines! Just don't forget to disable it afterwards.

File: docs/source/flow-overriding.rst [hunk 9]
```diff
@@ -0,0 +1,84 @@
+Configuring Responder Flows
+===========================
+
+A flow can be a fairly complex thing that interacts with many backend systems, and so it is likely that different users
+of a specific cordApp will require differences in how flows interact with their specific infrastructure.
+
+Corda supports this functionality by providing two mechanisms to modify your nodes' behaviour.
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: modify the behaviour of apps in your node.

File: docs/source/flow-overriding.rst [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+        @InitiatedBy(Initiator::class)
+        class SubResponder(otherSideSession: FlowSession) : BaseResponder(otherSideSession) {
+            override fun getMessage(): String {
+                return "This is the sub responder"
+            }
+        }
+
+   .. code-block:: kotlin
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: This code block is empty? Can we get a Java version of this snippet too?

File: docs/source/flow-overriding.rst [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+Corda would detect that both ``BaseResponder`` and ``SubResponder`` are configured for responding to ``Initiator``.
+Corda will then calculate the `hops` to ``FlowLogic`` and select the implementation which is furthest distance, ie: the most subclassed implementation.
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: why is hops in single backticks here?

File: docs/source/flow-overriding.rst [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+In the above example, ``SubResponder`` would be selected as the default responder for ``Initiator``
+
+.. note:: The flows do not need to be within the same cordApp, or package.
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: cordApp -> CorDapp

"or package, therefore to customise a shared app you obtained from a third party, you'd write your own CorDapp that subclasses the first."

File: docs/source/flow-overriding.rst [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+Overriding a Flow via Node Configuration
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: Flow -> flow
Node Configuration -> node configuration

File: docs/source/flow-overriding.rst [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+----------------------------------------
+
+Whilst the subclassing approach is likely to be useful for most applications, there is another mechanism to override this behaviour.
+This would be useful if for example, a specific cordApp user requires such a different responder that subclassing an existing flow
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: CorDapp

File: docs/source/flow-overriding.rst [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+would not be a good solution. In this case, it's possible to specify a hardcoded flow via the node configuration.
+The cordform plugin provides a ``flowOverride`` method within the ``deployNodes`` block which can be used to override a flow. In the below example, we will override
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: Cordform is useful for development and testing but surely this section of the docs is for node operators, who will just want to modify the config file directly?

File: docs/source/flow-overriding.rst [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+            rpcSettings {
+                address "localhost:10026"
+                adminAddress "localhost:10027"
+            }
+            extraConfig = ['h2Settings.address' : 'localhost:10035']
+            flowOverride("net.corda.Initiator", "net.corda.BaseResponder")
+        }
+
+
+If it is not possible to use the cordform plugin, the syntax for directly placing into the node config is as below
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: I'd put this section *first* as it's the most likely section to be needed, and mention the cordform/deploynodes equivalent later.

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+                RegisteredFlowContainer(
+                        initiatingFlowClass.java,
+                        initiatedFlowClass?.java,
+                        flowFactory,
+                        FlowType.CORE)
+        )
+        log.debug { "Installed core flow ${initiatingFlowClass.java.name}" }
+    }
+
+    //To verify the integrity of the current state, it is important that the tip of the responders is a unique weight
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: Comment style: `// Full sentences with punctuation and a space after the slashes.`

File: node/src/main/kotlin/net/corda/node/internal/FlowManager.kt [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+    private enum class FlowType {
+        CORE, CORDAPP
+    }
+
+    private data class RegisteredFlowContainer(val initiatingFlowClass: Class<out FlowLogic<*>>,
+                                       val initiatedFlowClass: Class<out FlowLogic<*>>?,
+                                       val flowFactory: InitiatedFlowFactory<FlowLogic<*>>,
+                                       val type: FlowType)
+
+    //this is used to sort the responding flows in order of "importance"
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/FlowManager.kt]: style

File: node/src/test/kotlin/net/corda/node/internal/NodeFlowManagerTest.kt [hunk 19]
```diff
@@ -0,0 +1,3 @@
+package net.corda.node.internal
+
+class NodeFlowManagerTest
```
Comment (mikehearn) [inline node/src/test/kotlin/net/corda/node/internal/NodeFlowManagerTest.kt]: This looks unused

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt (hunk 8 above)
Comment (mikehearn) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt]: This one still needs to be resolved - Kotlin warns on public functions that rely on return type inference these days because it can cause silent ABI breaks.

File: core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt [hunk 20]
```diff
@@ -151,11 +147,10 @@ class AttachmentSerializationTest {
     }
 
     private fun launchFlow(clientLogic: ClientLogic, rounds: Int, sendData: Boolean = false) {
-        server.registerFlowFactory(
-                ClientLogic::class.java,
-                InitiatedFlowFactory.Core { ServerLogic(it, sendData) },
-                ServerLogic::class.java,
-                track = false)
+        server.registerCordAppFlowFactory(
```
Comment (tlil) [inline core/src/test/kotlin/net/corda/core/serialization/AttachmentSerializationTest.kt]: Style, any reason why it can't be a single line? :)

File: docs/source/flow-overriding.rst [hunk 21]
```diff
@@ -0,0 +1,84 @@
+Configuring Responder Flows
+===========================
+
+A flow can be a fairly complex thing that interacts with many backend systems, and so it is likely that different users
+of a specific cordApp will require differences in how flows interact with their specific infrastructure.
```
Comment (tlil) [inline docs/source/flow-overriding.rst]: CorDapp

File: docs/source/flow-overriding.rst [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+            extraConfig = ['h2Settings.address' : 'localhost:10035']
+            flowOverride("net.corda.Initiator", "net.corda.BaseResponder")
+        }
+
+
+If it is not possible to use the cordform plugin, the syntax for directly placing into the node config is as below
+
+.. container:: codeset
+
+    .. code-block:: json
```
Comment (tlil) [inline docs/source/flow-overriding.rst]: `json` here and `groovy` further up. Does it render as expected?

File: core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt [hunk 23]
```diff
@@ -112,4 +101,23 @@ inline fun <reified R : Any> FlowLogic<*>.receiveAll(session: FlowSession, varar
 
 private fun Array<out Pair<FlowSession, Class<out Any>>>.enforceNoDuplicates() {
     require(this.size == this.toSet().size) { "A flow session can only appear once as argument." }
+}
+
+inline fun <reified P : FlowLogic<*>> TestStartedNode.registerCordAppFlowFactory(
```
Comment (tlil) [inline core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt]: We have `registerCordappFlows`, perhaps use `registerCordappFlowFactory` here for consistency?

File: docs/source/flow-overriding.rst (hunk 10 above)
Comment (roastario) [reply docs/source/flow-overriding.rst]: I do not know why, but if I remove that code block - nothing renders! 
Will try adding the java example in it.

File: docs/source/flow-overriding.rst (hunk 11 above)
Comment (roastario) [reply docs/source/flow-overriding.rst]: don't know what I was intending there - removed.

File: docs/source/flow-overriding.rst (hunk 22 above)
Comment (roastario) [reply docs/source/flow-overriding.rst]: yup, it's the only way to get sane formatting for the node.conf section.

File: testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt (hunk 8 above)
Comment (roastario) [reply testing/node-driver/src/main/kotlin/net/corda/testing/node/MockNetwork.kt]: Fixed, I have the type tips turned on and it looked like it was there!

File: docs/source/flow-overriding.rst (hunk 10 above)
Comment (roastario) [reply docs/source/flow-overriding.rst]: works nicely

File: docs/source/flow-overriding.rst (hunk 9 above)
Comment (roastario) [reply docs/source/flow-overriding.rst]: fixed.

File: docs/source/flow-overriding.rst (hunk 15 above)
Comment (roastario) [reply docs/source/flow-overriding.rst]: yep, agreed - reordered.

File: docs/source/flow-overriding.rst (hunk 16 above)
Comment (roastario) [reply docs/source/flow-overriding.rst]: yep, agreed - reordered.

File: node/src/test/kotlin/net/corda/node/internal/NodeFlowManagerTest.kt (hunk 19 above)
Comment (roastario) [reply node/src/test/kotlin/net/corda/node/internal/NodeFlowManagerTest.kt]: added the tests I meant to add!

File: core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt (hunk 23 above)
Comment (roastario) [reply core/src/test/kotlin/net/corda/core/flows/FlowTestsUtils.kt]: Yeap, changed to `TestStartedNode.registerCordappFlowFactory`

File: docs/source/flow-overriding.rst [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+
+Corda supports this functionality by providing two mechanisms to modify the behaviour of apps in your node.
+
+Subclassing a Flow
+------------------
+
+If you have a workflow which is mostly common, but also requires slight alterations in specific situations, most developers would be familiar
+with refactoring into `Base` and `Sub` classes. A simple example is shown below.
+
+java
```
Comment (mikehearn) [inline docs/source/flow-overriding.rst]: if you use the container/codeset thing, nice tabs should appear like for the other samples?
